### DISCUSSION TITLE
httpx giving UnsupportedProtocl error while running tests

### DISCUSSION THREAD

**Discussion by onlinejudge95** [encode/httpx#1643] [Potential Issue]
While implementing a unit test for a FastAPI server particularly in a GET route in which I am using `httpx` to send a get request(with some query params) to another service, I am receiving the following error
```bash
    @contextlib.contextmanager
    def map_httpcore_exceptions() -> typing.Iterator[None]:
        try:
            yield
        except Exception as exc:
            mapped_exc = None
    
            for from_exc, to_exc in HTTPCORE_EXC_MAP.items():
                if not isinstance(exc, from_exc):
                    continue
                # We want to map to the most specific exception we can find.
                # Eg if `exc` is an `httpcore.ReadTimeout`, we want to map to
                # `httpx.ReadTimeout`, not just `httpx.TimeoutException`.
                if mapped_exc is None or issubclass(to_exc, mapped_exc):
                    mapped_exc = to_exc
    
            if mapped_exc is None:  # pragma: nocover
                raise
    
            message = str(exc)
>           raise mapped_exc(message) from exc
E           httpx.UnsupportedProtocol: Unsupported URL protocol ''

../../../.pyenv/versions/3.9.4/envs/api-gateway/lib/python3.9/site-packages/httpx/_transports/default.py:78: UnsupportedProtocol
...
========================================================================================= short test summary info ==========================================================================================
FAILED tests/email/test_routes.py::test_fetch_email_fails_invalid_params[http://localhost:8000/api/email/-request_body0-400] - httpx.UnsupportedProtocol: Unsupported URL protocol ''
```
The test in question is
```python
@mark.parametrize("test_url,request_body,response_code", TEST_DATA_INVALID["GET"])
def test_fetch_email_fails_invalid_params(
    test_client, test_url, request_body, response_code
):
    response = test_client.get(test_url, params=request_body)

    assert response.status_code == response_code
    assert "request_id" not in response.json()
```
For full traceback please refer to the following [GitHub action](https://github.com/unofficialopensource-knit/api_gateway/pull/19/checks?check_run_id=2611371579) run
My entire codebase can be found [here](https://github.com/unofficialopensource-knit/api_gateway/tree/email-service-fetch-mails)
Link to the failing [test file](https://github.com/unofficialopensource-knit/api_gateway/blob/email-service-fetch-mails/tests/email/test_routes.py#L18)
Link to [System under test](https://github.com/unofficialopensource-knit/api_gateway/blob/email-service-fetch-mails/src/routers/email/routes.py#L67)

**Comment by lovelydinosaur** [comment]
You're probably not setting `EMAIL_MICROSERVICE_URL`, and when [this request gets made](https://github.com/unofficialopensource-knit/api_gateway/blob/email-service-fetch-mails/src/routers/email/routes.py#L78) it's to a URL like "/api/email".

We should probably include more information in that exception to make it clear what the full requested URL was, and make it more obvious that it was missing the "https://" or "http://" portion.

**Reply by onlinejudge95** [reply]
Makes sense as the other POST requests are blocked by FastAPI itself.
